File: Hard/longest_valid_parentheses.php
<?php 

class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function longestValidParentheses($s) {
        $countValidParentheses = 0;
        $storeIndex = array(-1);

        for ($i = 0; $i < strlen($s); $i++) {
            if ($s[$i] == "(") {
                array_push($storeIndex, $i);
            } else {
                array_pop($storeIndex);

                if(empty($storeIndex)) {
                    array_push($storeIndex, $i);
                } else {
                    $countValidParentheses =  max($countValidParentheses, $i - $storeIndex[count($storeIndex)-1]);
                }
            }
        }

        return $countValidParentheses;
    }
}

$solution = new Solution();

print_r($solution->longestValidParentheses("(()")); //2
print_r($solution->longestValidParentheses(")()());")); //4
print_r($solution->longestValidParentheses("")); //0
print_r($solution->longestValidParentheses("()(()")); //2
print_r($solution->longestValidParentheses("()(());")); //6
print_r($solution->longestValidParentheses("()")); //6

File: Hard/longest_valid_parentheses.go
package main

import "fmt"

func longestValidParentheses(s string) int {
	countValidParentheses := 0
	storeIndex := []int{-1}

	for i, c := range s {
		if string(c) == "(" {
			storeIndex = append(storeIndex, i)
		} else {
			storeIndex = storeIndex[:len(storeIndex)-1]
			if len(storeIndex) == 0 {
				storeIndex = append(storeIndex, i)
			} else {
				countValidParentheses = max(countValidParentheses, i-storeIndex[len(storeIndex)-1])
			}
		}
	}

	return countValidParentheses
}

func run_longestValidParentheses() {
	fmt.Println(longestValidParentheses("(()"))    //2
	fmt.Println(longestValidParentheses(")()())")) //4
	fmt.Println(longestValidParentheses(""))       //0
	fmt.Println(longestValidParentheses("()(()"))  //2
	fmt.Println(longestValidParentheses("()(())")) //6
	fmt.Println(longestValidParentheses("()"))     //2
}
